Advanced Template documentation (#2645)
tinymce/tinymce-docs: modules/ROOT/partials/configuration/advtemplate_list.adoc

[[advtemplate_list]]
== `+advtemplate_list+`

> This options allows for the loading of templates from external data sources. 

Reviewer: "options" to 'option"

> This options allows for the loading of templates from external data sources. 

Reviewer 2: options -> option

Reviewer 3: Urgent update, the `advtemplate_templates` options should not be mentioned in the public documentation. It is intended to use internally for testing or demo purposes. Once this option is set with a predefined constant template list, the plugin enables serverless mode, when all operations requests are redirected to the plugin's in-memory template database. It is important to understand that refreshing the page with the plugin in this mode resets the template list to the initial one. That is why using this option can mislead users and make them lose their data.

Author: To quote my commit message:

> 1. Removed `advtemplate_list.adoc ` from `/partials/configuration/`.
> 
> 2. Removed include statement pointing to `advtemplate_list.adoc ` from `advanced-templates.adoc`.
> 
> 3. Moved NOTE regarding unique id values requirement into `advanced-templates.adoc`.

> [[advtemplate_list]]
> == `+advtemplate_list+`
> 
> This options allows for the loading of templates from external data sources. 
> 
> To load templates from an external source, configure the `advtemplate_list` option to return a `Promise` containing the initial template list.
> 
> *Type:* `+Object+`
> 
> === Example: using `+advtemplate_list+` to load a template collection
> 
> [source,js]
> ----
> tinymce.init({
>   plugins: 'advtemplate',
>   toolbar: 'advtemplate',
>   advtemplate_list: () => {
>       return Promise.resolve([
>         {
>           id: 1,
>           title: 'Template 1'
>         },
>         {
>           id: 2,
>           title: 'Category 2',
>           items: [
>             {
>               id: 3,
>               title: 'Template 2.1',
>             },
>             {
>               id: 4,
>               title: 'Category 2.2',
>               items: []
>             }
>           ]
>         }
>     ]);
>   }
> })
> ----
> 
> NOTE: A unique id value must be provided for every input item.

Reviewer 4: ```suggestion
[[advtemplate_templates]]
== `+ advtemplate_templates +`

This options allows for the loading of templates from external data sources. 

To load templates from an external source, configure the `advtemplate_templates ` option to return a `Promise` containing the initial template list.

*Type:* `+Object+`

=== Example: using `+ advtemplate_templates +` to load a template collection

[source,js]
----
tinymce.init({
  plugins: 'advtemplate',
  toolbar: 'addtemplate inserttemplate',
  advtemplate_templates: () => {
      return Promise.resolve([
        {
          id: 1,
          title: 'Template 1'
        },
        {
          id: 2,
          title: 'Category 2',
          items: [
            {
              id: 3,
              title: 'Template 2.1',
            },
            {
              id: 4,
              title: 'Category 2.2',
              items: []
            }
          ]
        }
    ]);
  }
})
----

NOTE: A unique id value must be provided for every input item.
```
This option is called `advtemplate_templates`

Reviewer 3: @MitchCTiny Actually, your suggestion is not correct. 
The `advtemplate_templates` option should take a predefined list of templates (not a function), similar to the `templates` option in the old template plugin. The other options are meant for interacting with the external service. The `advtemplate_templates` option is only used to demonstrate the plugin's features with minimal configuration. Once the user sets it, the rest of the options are set automatically and communicate with in-memory storage. Therefore, after reloading the page, the template list is reset to the initial predefined list.

Reviewer 4: Right, I'll update the suggestion.

*Possible values:* see the example below

=== Example: using `advtemplate_list`

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  advtemplate_list: () => {
    return Promise.resolve([
      {
        id: '1',
        title: 'Resolving tickets',
        content: '<p>As we have not heard back from you in over a week, so we have gone ahead and resolved your ticket</p>'
      },
      {
        id: '2',
        title: 'Quick replies',
        items: [
          {
            id: '3',
            title: 'Message received',
            content: '<p>Just a quick note to say we have received your message, and will get back to you within 48 hours.</p>'
          },
          {
            id: '4',
            title: 'Progress update',
            content: '</p>Just a quick note to let you know we are still working on your case</p>'
          }
        ]
      }
    ]);
  }
});
----

The list returned by `advtemplate_list` must match the following requirements:

. Each list item must have a unique `id` value.
. Each list item must have a non-empty `title` value.
. Category items in the list must not contain nested categories.

> . Each template item in the list must have a non-empty `content` value.

Reviewer 3: ```suggestion
. Each template item in the list must have a non-empty `content` value.
```

> . Each template item in the list must have a non-empty `content` value.

Reviewer 3: The `content` property is not required in the template list item
```suggestion
```

[[advtemplate_list]]
== `advtemplate_list`

This option retrieves the list of available Templates.

*Type:* see the example below

*Possible values:* see the example below

=== Example: using `advtemplate_list`

[source,js]
----

> tinymce.init({
>   selector: 'textarea',  // change this value according to your html
>   advtemplate_list: () => {
>     return Promise.resolve(
>     [
>       {
>         id: '1',
>         title: 'Template 1'
>       },
>       {
>         id: '2',
>         title: 'Category 2',
>         items: [
>       {
>         id: '3',
>         title: 'Template 2.1',
>       },
>       {
>         id: '4',
>         title: 'Template 2.3',
>       }
>     ]
>   }
> ]
>     );
> });
> ----

Reviewer 4: ```suggestion
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  advtemplate_list: () => {
    return Promise.resolve(
    [
      {
        id: '1',
        title: 'Resolving tickets',
        content: '<p>As we have not heard back from you in over a week, so we have gone ahead and resolved your ticket</p>'
      },
      {
        id: '2',
        title: 'Quick replies',
        items: [
          {
            id: '3',
            title: 'Message received',
            content: '<p>Just a quick note to say we have received your message, and will get back to you within 48 hours.</p>'
          },
          {
            id: '4',
            title: 'Progress update',
            content: '</p>Just a quick note to let you know we are still working on your case</p>'
          }
        ]
      }
    ]
  );
});
----
```
Fixed indentation and added some sample content since 'content' is required

Reviewer 3: Actually the`content` is not required in this response. List response is used for creating the list of items for the Tree component. In order to keep traffic small the `content` property is removed from the list response

Author: > Actually the `content` is not required in this response.

A new [edit](https://github.com/tinymce/tinymce-docs/commit/f7f051e048e6300bb2f55cdcb81151f51b7fe996) that notes this.

> This example returns the following:
> 
> [source,js]
> ----
> [
>   {
>     id: '1',
>     title: 'Template 1'
>   },
>   {
>     id: '2',
>     title: 'Category 2',
>     items: [
>       {
>         id: '3',
>         title: 'Template 2.1',
>       },
>       {
>         id: '4',
>         title: 'Template 2.3',
>       }
>     ]
>   }
> ]
> ----

Reviewer 4: This is what will be shown in the dialog, the JSON is sent to the plugin to display.
```suggestion
```

[[advtemplate_list]]
== `advtemplate_list`

This option retrieves the list of available Templates.

*Type:* see the example below

*Possible values:* see the example below

=== Example: using `advtemplate_list`

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  advtemplate_list: () => {

>     return Promise.resolve(
>     [

Reviewer 4: ```suggestion
    return Promise.resolve([
```
Lining up indentation

      {
        id: '1',
        title: 'Resolving tickets',
        content: '<p>As we have not heard back from you in over a week, so we have gone ahead and resolved your ticket</p>'
      },
      {
        id: '2',
        title: 'Quick replies',
        items: [
          {
            id: '3',
            title: 'Message received',
            content: '<p>Just a quick note to say we have received your message, and will get back to you within 48 hours.</p>'
          },
          {
            id: '4',
            title: 'Progress update',
            content: '</p>Just a quick note to let you know we are still working on your case</p>'
          }
        ]
      }

>     ]
>   );
> });
> ----

Reviewer 4: ```suggestion
    ]);
  }
});
----
```
Just noticed a missing bracket
